File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/ChainOfResponsibility/concrete_handler.go
package ChainOfResponsibility

import "fmt"

type ConsoleLogger struct {
	*BaseLogger
}

func NewConsoleLogger(level int) *ConsoleLogger {
	c := ConsoleLogger{}
	c.BaseLogger = NewBaseLogger(level, &c)
	return &c
}

// Write 写入控制台日志
func (l *ConsoleLogger) Write(message string) {
	fmt.Println("Console: " + message)
}

type WarningLogger struct {
	*BaseLogger
}

func NewWarningLogger(level int) *WarningLogger {
	w := WarningLogger{}
	w.BaseLogger = NewBaseLogger(level, &w)
	return &w
}

// Write 写入警告日志
func (l *WarningLogger) Write(message string) {
	fmt.Println("Warning: " + message)
}

type ErrorLogger struct {
	*BaseLogger
}

func NewErrorLogger(level int) *ErrorLogger {
	e := ErrorLogger{}
	e.BaseLogger = NewBaseLogger(level, &e)
	return &e
}
func (l *ErrorLogger) Write(message string) {
	fmt.Println("Error: " + message)
}

File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/ChainOfResponsibility/ConcreteHandler.lua
-- 加载基础处理器
require("Handler")

-- ConsoleLogger 控制台日志处理器
--- @class ConsoleLogger : LoggerHandler
--- @field level number
--- @field nextLogger LoggerHandler
ConsoleLogger = {}
ConsoleLogger.__index = ConsoleLogger
setmetatable(ConsoleLogger, LoggerHandler)

-- 创建控制台日志处理器
function ConsoleLogger.new(level)
    local self = LoggerHandler.new(level)
    setmetatable(self, ConsoleLogger)
    return self
end
function ConsoleLogger:write(message)
    print("Standard Console: " .. message)
end

-- ErrorLogger 错误日志处理器
--- @class ErrorLogger : LoggerHandler
--- @field level number
--- @field nextLogger LoggerHandler
ErrorLogger = {}
ErrorLogger.__index = ErrorLogger
setmetatable(ErrorLogger, LoggerHandler)
function ErrorLogger.new(level)
    local self = LoggerHandler.new(level)
    setmetatable(self, ErrorLogger)
    return self
end
function ErrorLogger:write(message)
    print("Error Console: " .. message)
end

-- WarningLogger 警告日志处理器
--- @class WarningLogger : LoggerHandler
--- @field level number
--- @field nextLogger LoggerHandler
WarningLogger = {}
WarningLogger.__index = WarningLogger
setmetatable(WarningLogger, LoggerHandler)
function WarningLogger.new(level)
    local self = LoggerHandler.new(level)
    setmetatable(self, WarningLogger)
    return self
end
function WarningLogger:write(message)
    print("Warning Console: " .. message)
end